pixeltable/share/publish.py

```diff
@@ -165,45 +168,87 @@ def _download_bundle_from_s3(parsed_location: urllib.parse.ParseResult, bundle_f
     return bundle_path
 
 
-def _upload_bundle_with_presigned_url(bundle: Path, presigned_url: str, max_retries: int = 3) -> None:
-    file_size = bundle.stat().st_size
+def _create_retry_session(
+    max_retries: int = 3, backoff_factor: float = 1.0, status_forcelist: Optional[list] = None
+) -> requests.Session:
+    """Create a requests session with retry configuration"""
+    if status_forcelist is None:
+        status_forcelist = [
+            408,  # Request Timeout
+            429,  # Too Many Requests (rate limiting)
+            500,  # Internal Server Error (server-side error)
+            502,  # Bad Gateway (proxy/gateway got invalid response)
+            503,  # Service Unavailable (server overloaded or down)
+            504,  # Gateway Timeout (proxy/gateway timeout)
+        ]
+    retry_strategy = Retry(
+        total=max_retries,
+        read=max_retries,
+        connect=max_retries,
+        backoff_factor=backoff_factor,
+        status_forcelist=status_forcelist,
+        allowed_methods=['GET', 'PUT', 'POST', 'DELETE'],
+    )
+
+    session = requests.Session()
+    adapter = HTTPAdapter(max_retries=retry_strategy)
+    session.mount('https://', adapter)
+    return session
+
+
+def _upload_to_presigned_url(file_path: Path, url: str, max_retries: int = 3) -> requests.Response:
+    """Upload file with progress bar and retries"""
+    file_size = file_path.stat().st_size
+
     headers = {'Content-Length': str(file_size), 'Content-Type': 'application/octet-stream'}
-    for attempt in range(1, max_retries + 1):
-        try:
-            with open(bundle, 'rb') as f:
-                response = requests.put(presigned_url, data=f, headers=headers, stream=True, timeout=(60, 1800))
-                response.raise_for_status()
-                return
-        except (requests.exceptions.Timeout, requests.exceptions.SSLError, requests.exceptions.ConnectionError) as e:
-            if attempt == max_retries:
-                raise excs.Error(f'Upload failed after {max_retries} attempts: {e}') from e
-            time.sleep(1)  # wait before retrying
-            continue
-        except Exception as e:
-            raise excs.Error(f'Failed to upload : {e}') from e
-
-
-def _download_bundle_from_presigned_url(presigned_url: str, max_retries: int = 3) -> Path:
-    bundle_path = Path(Env.get().create_tmp_path())
-    for attempt in range(1, max_retries + 1):
-        try:
-            response = requests.get(presigned_url, stream=True, timeout=(60, 1800))
+
+    session = _create_retry_session(max_retries=max_retries)
+
+    try:
+        with (
+            open(file_path, 'rb') as f,
+            tqdm.wrapattr(f, 'read', total=file_size, desc='Uploading') as file_with_progress,
```

Comment:
Use the same tqdm configuration as for S3 transfers to ensure a consistent experience.